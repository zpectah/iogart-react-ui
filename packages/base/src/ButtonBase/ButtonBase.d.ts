import { ButtonBaseProps } from './types';

declare const ButtonBase: ButtonBaseProps;

export default ButtonBase;
