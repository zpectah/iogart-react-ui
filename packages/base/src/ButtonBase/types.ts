import {
    ReactNode,
    ButtonHTMLAttributes,
    AriaRole,
    FocusEventHandler,
    KeyboardEventHandler,
    MouseEventHandler,
    LegacyRef,
} from 'react';

import {
    IogartElementBase,
    iogartButtonElement,
    iogartAnchorElement,
} from '@iogart-react-ui/types/src';

export interface ButtonBaseProps<T = iogartButtonElement | iogartAnchorElement | HTMLElement> extends IogartElementBase {
    children?: ReactNode;
    elementType?: 'button' | 'a';
    disabled?: boolean;
    type?: ButtonHTMLAttributes<HTMLButtonElement>['type'],
    role?: AriaRole,
    onBlur?: FocusEventHandler;
    onFocus?: FocusEventHandler;
    onKeyDown?: KeyboardEventHandler;
    onKeyUp?: KeyboardEventHandler;
    onMouseEnter?: MouseEventHandler;
    onMouseDown?: MouseEventHandler;
    onMouseLeave?: MouseEventHandler;
    onMouseUp?: MouseEventHandler;
    onClick?: MouseEventHandler<T> | undefined;
    onDoubleClick?: MouseEventHandler<T> | undefined;
    ref?: LegacyRef<HTMLButtonElement> | undefined;
    href?: string;
    to?: string;
}

export interface useButtonBaseParameters extends ButtonBaseProps {}
export interface useButtonBaseReturn extends useButtonBaseParameters {}
