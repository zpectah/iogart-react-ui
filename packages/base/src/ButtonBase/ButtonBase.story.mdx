import { Meta, Story } from '@storybook/addon-docs';

import ButtonBase from './ButtonBase';
export const Template = (args) => <ButtonBase {...args} />;

<Meta
    title="Components/Action"
    component={ButtonBase}
/>

# Button Base
`@iogart-react-ui/base/ButtonBase`

```javascript
import { ButtonBase } from 'iogart-react-ui/base';
```

<Story
  name="ButtonBase"
  args={{
    children: 'Button Base',
  }}>
  {Template.bind({})}
</Story>
