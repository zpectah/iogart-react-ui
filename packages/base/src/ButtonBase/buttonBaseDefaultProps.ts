// TODO ???

const buttonBaseDefaultProps = {
    action: '',
    children: '',
    classes: '',
    className: '',
    component: 'a',
    disabled: '',
    href: '',
    onBlur: '',
    onClick: '',
    onDblClick: '',
    onContextMenu: '',
    onDragLeave: '',
    onFocus: '',
    onFocusVisible: '',
    onKeyDown: '',
    onKeyUp: '',
    onMouseDown: '',
    onMouseLeave: '',
    onMouseUp: '',
    onTouchEnd: '',
    onTouchMove: '',
    onTouchStart: '',
    sx: '',
    tabIndex: '',
    type: '',
};

export default buttonBaseDefaultProps;
