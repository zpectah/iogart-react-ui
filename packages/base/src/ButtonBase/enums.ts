export enum ButtonBaseElementKeys {
    button = 'button',
    a = 'a',
}
