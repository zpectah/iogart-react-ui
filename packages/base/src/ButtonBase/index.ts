export { default as ButtonBase } from './ButtonBase';
export { default as useButtonBase } from './useButtonBase';
export * from './types';
export * from './ButtonBase.d';
