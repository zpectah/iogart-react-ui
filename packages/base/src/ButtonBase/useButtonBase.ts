import { useRef } from 'react';

import { useButtonBaseParameters, useButtonBaseReturn } from './types';

const useButtonBase = (props: useButtonBaseParameters) => {
    const { ref, className, ...rest } = props;

    const elementRef = useRef<HTMLButtonElement | HTMLAnchorElement | HTMLElement>();

    const getClassName = () => {

        return `iogart-base iogart-button-base${className && ` ${className}`}`;
    };

    return {
        ref: ref || elementRef,
        className: getClassName(),
        ...rest,
    } as useButtonBaseReturn;
};

export default useButtonBase;
