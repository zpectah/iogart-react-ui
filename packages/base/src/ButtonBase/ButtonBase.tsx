import { FC, createElement } from 'react';

import { ButtonBaseElementKeys } from './enums';
import { ButtonBaseProps } from './types';
import useButtonBase from './useButtonBase';

const ButtonBase: FC<ButtonBaseProps> = (props) => {
    const {
        children,
        elementType = ButtonBaseElementKeys['button'],
        ...rest
    } = props;

    const updatedProps = useButtonBase({ ...rest });

    return createElement(
        elementType,
        updatedProps,
        children,
    );
};

export default ButtonBase;
