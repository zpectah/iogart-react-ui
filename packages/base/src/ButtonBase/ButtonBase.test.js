import renderer from 'react-test-renderer';

import ButtonBase from './ButtonBase';

it('Button Base - onMouseEnter & onMouseLeave', () => {

    const component = renderer.create(
        <ButtonBase>Button base</ButtonBase>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    renderer.act(() => {
        tree.props.onMouseEnter();
    });

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    renderer.act(() => {
       tree.props.onMouseLeave();
    });

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

});
