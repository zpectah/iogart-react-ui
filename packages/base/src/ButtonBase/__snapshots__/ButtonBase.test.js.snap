// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Base - onMouseEnter & onMouseLeave 1`] = `
<button
  className="iogart-ui iogart-base iogart-button-base"
  onBlur={[Function]}
  onClick={[Function]}
  onDoubleClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  type="button"
>
  Button base
</button>
`;

exports[`Button Base - onMouseEnter & onMouseLeave 2`] = `
<button
  className="iogart-ui iogart-base iogart-button-base iogart--hover"
  onBlur={[Function]}
  onClick={[Function]}
  onDoubleClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  type="button"
>
  Button base
</button>
`;

exports[`Button Base - onMouseEnter & onMouseLeave 3`] = `
<button
  className="iogart-ui iogart-base iogart-button-base"
  onBlur={[Function]}
  onClick={[Function]}
  onDoubleClick={[Function]}
  onFocus={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  type="button"
>
  Button base
</button>
`;
